src/k2/module/k23si/Module.cpp

```diff
@@ -1219,42 +1233,78 @@ K23SIPartitionModule::handleTxnFinalize(dto::K23SITxnFinalizeRequest&& request)
         return RPCResponse(dto::K23SIStatus::OperationNotAllowed("cannot commit missing key"), dto::K23SITxnFinalizeResponse());
     }
 
-    // we found a record from this transaction
-    switch(rec->status) {
-        case dto::DataRecord::WriteIntent: {
-            // if it is currently a write intent, modify as needed
-            if (request.action == dto::EndAction::Commit) {
-                K2LOG_D(log::skvsvr, "Partition: {}, committing {}, in txn {}", _partition, request.key, txnId);
-                rec->status = dto::DataRecord::Committed;
-            }
-            else {
-                K2LOG_D(log::skvsvr, "Partition: {}, aborting {}, in txn {}", _partition, request.key, txnId);
-                rec->status = dto::DataRecord::Aborted;
+    // TODO-persistence We should handle the cases when the record is updated in-memory but not persisted yet
+    VersionSet& versions = IndexerIt->second;
+    bool found = false;
+    bool isWI = false;
+    bool isAborted = false;
+
+    // Check for matching WI first
+    if (versions.WI.has_value() && versions.WI->txnId == txnId) {
+        found = true;
+        isWI = true;
+        isAborted = versions.WI->status == dto::WriteIntent::Status::Aborted;
+    }
+    // Check for a matching committed record if needed
+    if (!found) {
+        for (const CommittedRecord& record : versions.committed) {
+            // timestamps are unique, so they can identify a txn match
+            if (record.timestamp.compareCertain(txnId.mtr.timestamp) == 0) {
+                found = true;
             }
-            break;
         }
-        case dto::DataRecord::Committed:
-            // don't trigger the failure response if the action matches the state
-            if (request.action == dto::EndAction::Commit) break;
-            K2LOG_D(log::skvsvr, "Partition: {}, cannot abort committed record: {}", _partition, *rec);
-            return RPCResponse(dto::K23SIStatus::OperationNotAllowed("cannot finalize txn"), dto::K23SITxnFinalizeResponse());
-        case dto::DataRecord::Aborted:
-            // don't trigger the failure response if the action matches the state
-            if (request.action == dto::EndAction::Abort) break;
-            K2LOG_D(log::skvsvr, "Partition: {}, cannot commit aborted record: {}", _partition, *rec);
-            return RPCResponse(dto::K23SIStatus::OperationNotAllowed("cannot finalize txn"), dto::K23SITxnFinalizeResponse());
-        default:
-            // the action did not match the state
-            K2LOG_D(log::skvsvr,
-                "Partition: {}, failing finalize due to action mismatch {}, in txn {}, have status={}, asked={}",
-                _partition, request.key, txnId, rec->status, request.action);
+    }
+
+    if (request.action == dto::EndAction::Abort) {
```

Comment:
I think this block would read better if you used a switch instead of another if-elseif layer